### PR TITLE
introduce the design spec for implementation of model reader writer interfaces in the generator

### PR DESCRIPTION
**Pull request by ArcturusZhang**
# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: docs/design/model-reader-writer-implementation-design.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+}
+```
+
+The default behavior of the most widely used JSON serializer and deserializer is usually: they serialize classes and structs in the same way, assuming that they are models containing properties. This is true for our classes, but for structs, it is not always true. The extensible enums are modeled as structs, but they are actually serialized to a string value in our internal implementation when building the payload or parsing the response. For example, the `Foo` model above are expected to be serialized to the following JSON:
+```json
+{
+    "bar": "value"
+}
+```
+but in the default behavior of `System.Text.JsonSerializer`, it will be serialized to:
```
Comment (ArthurMa1978) [inline docs/design/model-reader-writer-implementation-design.md:45]: We need mention that the `ModelReaderWriter` does not support `System.Text.JsonSerializer` now, so even we implement it in .Net SDK but still can't serialize the model properly by using `System.Text.JsonSerializer`.

**Review by ArthurMa1978 — APPROVED** [review]

File: docs/design/model-reader-writer-implementation-design.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+If the model supports `xml` format, it will contain the following methods:
+1. `IXmlSerializable.Write(XmlWriter writer, string nameHint)` method.
+2. `WriteInternal(XmlWriter writer, string nameHint, ModelReaderWriterOptions options)` method.
+3. the static `Deserialize{ModelName}(XElement element, ModelReaderWriterOptions options = null)` method.
+
+#### The changes on models
+
+All models will now have a field `private IDictionary<string, BinaryData> _serializedAdditionalRawData`, if the model has a base type, the field should change to `private protected IDictionary<string, BinaryData> _serializedAdditionalRawData`.
+
+This private field works exactly the same as additional properties property, it should vanish when the model or any of its base type has an additional properties property defined.
```
Comment (lirenhe) [inline docs/design/model-reader-writer-implementation-design.md:123]: What if the base type has an additional properties defined as:
```
public IDictionary<string, string> AdditionalProperties { get; }  ?
```
Will we convert the type into  
```
private IDictionary<string, BinaryData>?
```

Comment (ArcturusZhang) [reply docs/design/model-reader-writer-implementation-design.md:127]: this is indeed an issue. We need to resolve this later.
I will create an issue for this.

Comment (ArcturusZhang) [reply docs/design/model-reader-writer-implementation-design.md:127]: Issue created: https://github.com/Azure/autorest.csharp/issues/4128

File: docs/design/model-reader-writer-implementation-design.md (hunk 1 above)
Comment (ArcturusZhang) [reply docs/design/model-reader-writer-implementation-design.md:45]: added a paragraph about this

**Pull request merged**